pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -71,6 +71,7 @@ type BindMechanism interface {
 
 type podNICImpl struct {
 	cacheFactory cache.InterfaceCacheFactory
+	handler      NetworkHandler
```

Comment:
Can you please explain what do we gain by having this change?
You wrote -
>> enables us to properly mock it in the unit tests
I couldn't see any special change in the unit tests.

If we want to get rid of the `Handler` global, I would change the `Handler` to be a singleton.

Reply:
> Can you please explain what do we gain by having this change?
> You wrote -
> 
> > enables us to properly mock it in the unit tests
> 
> I couldn't see any special change in the unit tests.
Now that handler is a clear dependency of podnic we just pass a mock to the dependency so we don't have to change a global or rewrite a singleton initialize functioon or the like.

```golang
mockNetwork = NewMockNetworkHandler(ctrl)                               
        podnic = podNICImpl{cacheFactory: cacheFactory, handler: mockNetwork}   
```

> 
> If we want to get rid of the `Handler` global, I would change the `Handler` to be a singleton.

ditto https://github.com/kubevirt/kubevirt/pull/4977#discussion_r618289722

Reply:
`common.go` is a utility class with helper methods. The only reason we have the `NetworkHandler` is because of our tests limitations to mock global methods.

I find it very confusing to pass a struct that has no state and is there just because of a tests limitation as a parameter to so many methods. It gives the feeling that the may be several implementations of the `handler`. I find it very confusing.

So yes, globals most of the times are a bad practice. But in this case, IMO, the proposed alternative is not better.

Reply:
@AlonaKaplan , my 2 cents:

> > `common.go` is a utility class with helper methods. The only reason we have the `NetworkHandler` is because of our tests limitations to mock global methods.
> 
> also it decouples low level primitives, for example if we want to use different library than "netlink" it will help with that.
> 
> > I find it very confusing to pass a struct that has no state and is there just because of a tests limitation as a parameter to so many methods.

This struct already exists. I think @qinqon is just making it explicit, instead of keeping it hidden behind smoke and mirrors.

> 
> Only `createAndBindTapToBridge` receive the handler at the function, the rest use it from the podnic or from the binding struct.
> 
> ```
> podinterface.go:        err := createAndBindTapToBridge(b.handler, tapDeviceName, b.bridgeInterfaceName, b.queueCount, *b.launcherPID, int(b.vif.Mtu), "0")
> ```

Not sure this would help, but you could take this refactor a step further and add the `createAndBindTapToBridge` to the bidge / masquerade struct implementations and make it a method - that way, you'd not need to inject the network driver into it.

> 
> > It gives the feeling that the may be several implementations of the `handler`. I find it very confusing.

The feeling that there might be several implementations of the handler should come from the fact that the handler is an interface. Which it already is - and has been from the start.

> 
> There has being always two implementations the real one and the mock one, and hiding this with a global var make thing even more confusing, if we think about the issue we have at this PR trying to find why podNICFactory was needed podinterface_test.go it illustrate the problem I think.
> 
> > So yes, globals most of the times are a bad practice. But in this case, IMO, the proposed alternative is not better.
> 
> I understand that looks a little too much to promote handler as podnic/bindmechanism member just for unit test, but in fact what we are doing here is to make clear the separation we were hidding with global vars, and IMO that's always a good thing.

+1